### ISSUE TITLE
KoboldCpp accessibility issues with NVDA screen reader

### ISSUE THREAD

**Issue by inspector3535** [LostRuins/koboldcpp#1661]
greetings;
I am writing to report some accessibility issues I have encountered while using Koboldcpp with the NVDA (NonVisual Desktop Access) screen reader. As a visually impaired user, I rely on screen readers to interact with software, and I've found Koboldcpp's interface to be currently inaccessible.
Specifically, I have observed the following problems:
• Missing Product Version Information: NVDA reports an error indicating that the ProductVersion information is missing from the Koboldcpp.exe file. This lack of metadata can hinder NVDA's ability to properly identify and interact with the application.
• Unnamed and Unknown Interface Elements: Many interface elements lack proper names, roles, and descriptions that can be announced by a screen reader. NVDA identifies elements as "None" or "UNKNOWN," making it impossible to understand their function or purpose.
• Tkinter Accessibility: I understand Koboldcpp is built with Tkinter, and it's possible the accessibility limitations of Tkinter are affecting the usability of the program with a screen reader. IAccessible elements like "accName", "accDescription" and "accValue" are also showing up as "None".
below I am attaching a snippet of the NVDA log file which contains relevant information that may help you identify the source of the problem.
I would greatly appreciate it if you could address these accessibility issues in future updates. Properly labeling interface elements, providing descriptive text, and ensuring compatibility with screen readers would significantly improve the usability of Koboldcpp for visually impaired users like myself.
Furthermore, I would be happy to volunteer my time to assist with accessibility testing or provide feedback during the development process, should you require it. I can offer insight from a screen reader user perspective.
Thank you for your time and consideration. I look forward to seeing improved accessibility in future versions of Koboldcpp.
Sincerely,
Mert Anlar
Email: mertcan9507@gmail.com
```
NVDA Version: 2025.1.2 (2025.1.2.36913)
OS: Windows 10 build 22H2 (19045.6093)
WARNING - fileUtils.getFileVersionInfo (12:38:58.955) - MainThread (9956):
Invalid or unavailable version info attribute for 'C:\\Users\\Administrator\\Downloads\\koboldcpp.exe': ProductVersion
INFO - globalCommands.script_navigatorObject_devInfo (12:38:59.156) - MainThread (9956):
Developer info for navigator object:
name: None
role: UNKNOWN
processID: 10596
roleText: None
states: FOCUSABLE, FOCUSED
isFocusable: True
hasFocus: True
Python object: <NVDAObjects.Dynamic_ClipboardCommandAnnouncementContentGenericClientIAccessibleWindowNVDAObject object at 0x008442B0>
Python class mro: (<class 'NVDAObjects.Dynamic_ClipboardCommandAnnouncementContentGenericClientIAccessibleWindowNVDAObject'>, <class 'globalPlugins.NVDAExtensionGlobalPlugin.clipboardCommandAnnouncement.ClipboardCommandAnnouncement'>, <class 'NVDAObjects.IAccessible.ContentGenericClient'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'documentBase.TextContainerObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <class 'garbageHandler.TrackedObject'>, <class 'object'>)
description: None
location: RectLTWH(left=0, top=23, width=1920, height=769)
value: ''
TextInfo: <class 'displayModel.DisplayModelTextInfo'>
appModule: AppModule(appModuleHandler, appName='koboldcpp', processID=10596)
appModule.productName: 'koboldcpp'
appModule.productVersion: None
appModule.helperLocalBindingHandle: c_long(200655648)
appModule.appArchitecture: 'AMD64'
windowHandle: 7145002
windowClassName: 'TkChild'
windowControlID: 0
windowStyle: 1442840576
extendedWindowStyle: 4
windowThreadID: 1896
windowText: ''
displayText: ''
IAccessibleObject: <POINTER(IAccessible) ptr=0xc0ed4c0 at a285df0>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=7145002, objectID=-4, childID=0
IAccessible accName: None
IAccessible accRole: ROLE_SYSTEM_CLIENT
IAccessible accState: STATE_SYSTEM_FOCUSED, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1048580)
IAccessible accDescription: None
IAccessible accValue: None
```

**Comment by LostRuins** [comment]
Unfortunately, KoboldCpp does use CustomTkinter for it's GUI which does not support aria labels and generally has poor screen reader accessibility.

However there is a pretty good solution - simply use KoboldCpp from the command line! All KoboldCpp features are accessible via the CLI and support pure non-visual, text only inputs.

To learn the commands, simply run `koboldcpp.exe --help`

You can launch a model with a command like `koboldcpp.exe --model yourmodel.gguf --usecuda --gpulayers 20` for example. All commands are documented in help and provide the parameters that can be used.

**Comment by henk717** [comment]
As lostruins said for screen reader users its likely easier to skip the launcher entirely as TK will limit what we can do there. Once you are launched the webui that followed should be screen reader friendly.

In addition I currently have the assumption that commandline launching is more user friendly for screen reader users than having to navigate a launcher UI in general. Am I correct in this?

**Comment by henk717** [comment]
Also worth pointing out, KoboldCpp supports launching configuration files. We can make those on request (and you can also make them using --exportconfig filename along with the settings you want.

So if you need a more specialized config and find it hard to use the command line parametets we can accomodate by saving it for you.
